component/loki/source/gelf/component.go

```diff
@@ -0,0 +1,141 @@
+package gelf
+
+import (
+	"context"
+	"sync"
+
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/component/common/loki"
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/clients/pkg/promtail/scrapeconfig"
+	"github.com/grafana/loki/clients/pkg/promtail/targets/gelf"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name: "loki.source.gelf",
+		Args: Arguments{},
+
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+var _ component.Component = (*Component)(nil)
+
+// Component is a receiver for graylog formatted log files.
+type Component struct {
+	mut       sync.RWMutex
+	target    *gelf.Target
+	o         component.Options
+	metrics   *gelf.Metrics
+	handler   *handler
+	receivers []loki.LogsReceiver
+}
+
+// Run starts the component.
+func (c *Component) Run(ctx context.Context) error {
+	defer func() {
+		c.target.Stop()
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.handler.c:
+			c.mut.RLock()
+			lokiEntry := loki.Entry{
+				Labels: entry.Labels,
+				Entry:  entry.Entry,
+			}
+			lokiEntry.Labels["job"] = model.LabelValue(c.o.ID)
+			for _, r := range c.receivers {
+				r <- lokiEntry
+			}
+			c.mut.RUnlock()
+		}
+	}
+}
+
+// Update updates the fields of the component.
+func (c *Component) Update(args component.Arguments) error {
+	newArgs := args.(Arguments)
+
+	c.mut.Lock()
+	defer c.mut.Unlock()
+
+	if c.target != nil {
+		c.target.Stop()
+	}
+	c.receivers = newArgs.Receivers
+	t, err := gelf.NewTarget(c.metrics, c.o.Logger, c.handler, nil, convertConfig(newArgs))
+	if err != nil {
+		return err
+	}
+	c.target = t
+	return nil
+}
+
+// Arguments are the arguments for the component.
+type Arguments struct {
+	// ListenAddress only supports UDP.
+	ListenAddress        string              `river:"listen_address,attr,optional"`
+	UseIncomingTimestamp bool                `river:"use_incoming_timestamp,attr,optional"`
+	Receivers            []loki.LogsReceiver `river:"forward_to,attr"`
```

Comment:
This is a question for other `loki.source.*` components as well, so looking for a temperature check.

Do we allow each component to only define a single target? Or should we do what scrape_configs do and wrap these into a block so the user can define as many as they'd want under a single component?

After doing the block thing with the `syslog` target, I think that on one hand, I like the simplicity that this brings, plus having them as separate graph nodes, but not sure what's best from a UX perspective.

cc @mattdurham @rfratto

Reply:
I like 1 per component; there *might* be exceptions, but this makes it cleaner and the config much cleaner.

Reply:
What does "target" mean in this case? Server?

Reply:
Yes, that's the idea. After catching up with Matt and arrived at the same conclusion.

We'll have sources like this or the Heroku only allow launching a single listener, and others like GCPLog that might need to read from multiple subscriptions or Syslog that allows both TCP and UDP allow defining multiple listeners.

> for the above example: at most one TCP listener AND at most one UDP listener, up to two total listeners

This is _not_ implemented currently in the Syslog receiver, but we can think about this restriction. I'm not sure if in the Syslog world it's valid to have multiple application log on different ports, but I'll check it out.